README.md
nebenan-eventproxy
====================

Nebenan eventproxy helper library


CHANGELOG.md
## v1.1.0 
- Now with typescript

## v1.0.0 Initial release
- Package extracted from `nebenan-helper`


src/index.test.js
const { assert } = require('chai');
const sinon = require('sinon');
const proxyquire = require('proxyquire').noPreserveCache();

let eventproxy;
let handler;
let mainNodeHandler;
let clock;

const HANDLER_ATTACH_DELAY = 101;

describe('eventproxy', () => {
  beforeEach(() => {
    global.window = true;

    const mainNode = {};
    mainNode.addEventListener = sinon.spy(
      (event, callback) => { mainNodeHandler = callback; },
    );
    mainNode.removeEventListener = sinon.spy();

    global.document = {};
    global.document.addEventListener = sinon.spy((event, callback) => { handler = callback; });
    global.document.removeEventListener = sinon.spy();
    global.document.querySelector = () => mainNode;

    global.addEventListener = sinon.spy((event, callback) => { handler = callback; });
    global.removeEventListener = sinon.spy();

    eventproxy = proxyquire('./index', {});

    clock = sinon.useFakeTimers();
  });

  after(() => {
    delete global.window;
    delete global.document;
    delete global.addEventListener;
    delete global.removeEventListener;
  });

  afterEach(() => {
    clock.restore();
  });

  it('server safe', () => {
    delete global.window;
    global.document.addEventListener = sinon.spy();
    const servereventproxy = require('./index');
    const spy = sinon.spy();

    const unsubscribe = servereventproxy('diceroll', spy);

    assert.isTrue(global.document.addEventListener.notCalled, 'didn\'t initialize on server by default');
    assert.isTrue(spy.notCalled, 'didn\'t call callback');
    assert.isFunction(unsubscribe, 'returns a function');

    unsubscribe();
  });

  it('creates listener', () => {
    const spy = sinon.spy();
    const unsubscribe = eventproxy('diceroll', spy);
    clock.tick(HANDLER_ATTACH_DELAY);
    assert.isTrue(global.document.addEventListener.calledOnce, 'initialized on client');
    assert.isTrue(spy.notCalled, 'didn\'t call callback');
    unsubscribe();
  });

  it('does not fire handler during threshold', () => {
    const spy = sinon.spy();
    const unsubscribe = eventproxy('test', spy);
    handler({ type: 'test' });
    clock.tick(HANDLER_ATTACH_DELAY);
    handler({ type: 'test' });

    assert.isTrue(spy.calledOnce);
    unsubscribe();
  });

  it('handles events correctly', () => {
    const spy = sinon.spy();
    const spy2 = sinon.spy();

    const unsubscribe = eventproxy('test', spy);
    const unsubscribe2 = eventproxy('test2', spy2);
    clock.tick(HANDLER_ATTACH_DELAY);

    handler({ type: 'test' });
    handler({ type: 'test' });
    unsubscribe();

    handler({ type: 'test' });
    handler({ type: 'test2' });
    unsubscribe2();

    assert.isTrue(spy.calledTwice, 'called twice for correct event');
    assert.isTrue(spy2.calledOnce, 'called once for correct event 2');
  });

  it('throttles scroll events', () => {
    const spy = sinon.spy();

    const unsubscribe = eventproxy('scroll', spy);
    clock.tick(HANDLER_ATTACH_DELAY);
    handler({ type: 'scroll' });
    handler({ type: 'scroll' });
    handler({ type: 'scroll' });
    handler({ type: 'scroll' });
    handler({ type: 'scroll' });
    unsubscribe();

    assert.isTrue(spy.calledOnce, 'called once for correct event');
  });

  it('attaches click events to #main', () => {
    const spy = sinon.spy();
    eventproxy('click', spy);
    clock.tick(HANDLER_ATTACH_DELAY);

    mainNodeHandler({ type: 'click' });

    assert.isTrue(spy.calledOnce, 'handler called for event on correct element');
  });

  it('multiple unsubscribes', () => {
    const spy = sinon.spy();

    const unsubscribe = eventproxy('test', spy);
    clock.tick(HANDLER_ATTACH_DELAY);

    handler({ type: 'test' });
    unsubscribe();
    unsubscribe();
    unsubscribe();
    handler({ type: 'test' });

    assert.isTrue(spy.calledOnce, 'called once for correct event');
  });

  it('listener unsubscribes another listener', () => {
    let unsubscribe;
    const spy = sinon.spy(() => unsubscribe && unsubscribe());

    unsubscribe = eventproxy('test', spy);
    unsubscribe = eventproxy('test', spy);
    clock.tick(HANDLER_ATTACH_DELAY);

    handler({ type: 'test' });

    assert.isTrue(spy.calledOnce, 'call count correct');
  });
});


src/index.js
import debounce from 'lodash/debounce';
import throttle from 'lodash/throttle';

const RESIZE_RATE = 300;
const SCROLL_RATE = 100;
const HANDLER_CALL_DELAY = 100;

// ========================================================================================
// Initialization
// ========================================================================================
const isDOMAvailable = typeof window !== 'undefined';

const settingsMap = {};
const eventMap = {};
const noop = () => {};
const defaultSettings = {};


const createEventSettings = () => {
  settingsMap.resize = {
    emitter: global,
    wrapper(callback) { return debounce(callback, RESIZE_RATE); },
  };

  settingsMap.scroll = {
    emitter: global,
    wrapper(callback) { return throttle(callback, SCROLL_RATE); },
  };

  // React 17 changed where it attaches it's event listeners.
  // React 16 attached all event listeners to document. React 17 attaches all event
  // listeners to the react root node.
  // Given an event handler of type 'click' attaches a global click handler onto
  // document, the click event would bubble up to the newly attached document click
  // handler (which is, in most cases, an unexpected bahvior). We can prevent this
  // from attaching specific events to the react root node instead.
  settingsMap.click = {
    emitter: global.document.querySelector('#main') || global.document,
  };

  defaultSettings.emitter = global.document;
};

// ========================================================================================
// Utility functions
// ========================================================================================
const getEventData = (event) => {
  if (!eventMap[event]) eventMap[event] = { listeners: {}, lastIndex: 0, listenersLength: 0 };
  return eventMap[event];
};

const getEventSettings = (event) => settingsMap[event] || defaultSettings;

// # This is a very dirty fix to a bad problem.
// When attaching listeners of a given event type DURING the execution of an event handler
// of this type, it needs to be prevented that the newly attached listener gets called immediately.
// It is not a trivial task to get some code to execute after all event handlers for a
// DOM element have been called. Some browser engines schedule tasks differently. Checking
// for time is the most simple fix to the problem.
//
// How to prevent: Do not attach event listeners in event handlers.
// https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/
const isReadyForExecution = (handler) => (
  handler && Date.now() - handler._attachedAt > HANDLER_CALL_DELAY
);

const handleEmitterEvent = (event) => {
  const eventData = getEventData(event.type);
  // Item may have been deleted during iteration cycle
  Object.keys(eventData.listeners).forEach((id) => {
    const handler = eventData.listeners[id];
    if (isReadyForExecution(handler)) handler(event);
  });
};

const attachEmitterHandler = (event, eventData, eventSettings) => {
  if (eventData.listenersLength === 1) {
    eventSettings.emitter.addEventListener(event, handleEmitterEvent, { passive: true });
  }
};

const detachEmitterHandler = (event, eventData, eventSettings) => {
  if (eventData.listenersLength === 0) {
    eventSettings.emitter.removeEventListener(event, handleEmitterEvent, { passive: true });
  }
};

// ========================================================================================
// Public api
// ========================================================================================
const removeListener = (event, id) => {
  const eventData = getEventData(event);
  const eventSettings = getEventSettings(event);
  // protect from being called twice
  if (!eventData.listeners[id]) return;

  delete eventData.listeners[id];
  eventData.listenersLength -= 1;

  detachEmitterHandler(event, eventData, eventSettings);
};

/**
 * @function
 * @param {string} event
 * @param {function} callback
 * @return {removeEventListener|noop}
 */
const addListener = (event, callback) => {
  if (typeof event !== 'string') throw new Error('Event name required');
  if (typeof callback !== 'function') throw new Error('Listener function required');
  if (!isDOMAvailable) return noop;

  const eventData = getEventData(event);
  const eventSettings = getEventSettings(event);

  eventData.lastIndex += 1;
  const id = eventData.lastIndex;
  const handler = eventSettings.wrapper ? eventSettings.wrapper(callback) : callback;
  handler._attachedAt = Date.now();

  eventData.listeners[id] = handler;
  eventData.listenersLength += 1;

  attachEmitterHandler(event, eventData, eventSettings);

  const removeEventListener = () => {
    if (typeof handler.cancel === 'function') {
      handler.cancel();
    }

    removeListener(event, id);
  };
  // Allow to empty the calls queue
  removeEventListener.cancel = handler.cancel;

  return removeEventListener;
};

if (isDOMAvailable) createEventSettings();

export default addListener;


src/index.ts
import debounce from 'lodash/debounce';
import throttle from 'lodash/throttle';

type EventType = keyof GlobalEventHandlersEventMap;

type EventData = {
  listeners: Record<string, HandlerFunction>;
  lastIndex: number;
  listenersLength: number;
};

type CallbackFunction = (event:Event) => void;

type EventSettings = {
  emitter?: Document | typeof globalThis;
  wrapper?: (callback: CallbackFunction) => Function;
};

type HandlerFunction = CallbackFunction & { _attachedAt: number, cancel?:Function };

const RESIZE_RATE = 300;
const SCROLL_RATE = 100;
const HANDLER_CALL_DELAY = 100;

const isDOMAvailable = typeof window !== 'undefined';

const settingsMap: Record<string, EventSettings> = {};
const eventMap: Record<string, EventData> = {};
const noop = () => {};
const defaultSettings: EventSettings = {};

const createEventSettings = () => {
  settingsMap.resize = {
    emitter: globalThis,
    wrapper(callback) { return debounce(callback, RESIZE_RATE); },
  };

  settingsMap.scroll = {
    emitter: globalThis,
    wrapper(callback) { return throttle(callback, SCROLL_RATE); },
  };

  // React 17 changed where it attaches its event listeners.
  // React 16 attached all event listeners to document. React 17 attaches all event
  // listeners to the React root node.
  // Given an event handler of type 'click' attaches a global click handler onto
  // document, the click event would bubble up to the newly attached document click
  // handler (which is, in most cases, an unexpected behaviour). We can prevent this
  // from attaching specific events to the React root node instead.
  settingsMap.click = {
    // TODO make selector configurable, see
    // https://github.com/good-hood-gmbh/nebenan-frontend/pull/1899#discussion_r1150820056
    // @ts-ignore
    emitter: globalThis.document.querySelector('#main') || globalThis.document,
  };

  defaultSettings.emitter = globalThis.document;
};

// ========================================================================================
// Utility functions
// ========================================================================================
function getEventData<TType extends EventType>(eventType:TType): EventData {
  if (!eventMap[eventType]) eventMap[eventType] = { listeners: {}, lastIndex: 0, listenersLength: 0 };
  return eventMap[eventType];
}

function getEventSettings<TType extends EventType>(eventType:TType): EventSettings {
  return settingsMap[eventType] || defaultSettings;
}

// # This is a very dirty fix to a bad problem.
// When attaching listeners of a given event type DURING the execution of an event handler
// of this type, it needs to be prevented that the newly attached listener gets called immediately.
// It is not a trivial task to get some code to execute after all event handlers for a
// DOM element have been called. Some browser engines schedule tasks differently. Checking
// for time is the simplest fix to the problem.
//
// How to prevent: Do not attach event listeners in event handlers.
// https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/
const isReadyForExecution = (handler: Function & { _attachedAt: number }): boolean => (
  handler && Date.now() - handler._attachedAt > HANDLER_CALL_DELAY
);

const handleEmitterEvent = (event: Event) => {
  const eventData = getEventData(event.type as EventType);
  // Item may have been deleted during iteration cycle
  Object.keys(eventData.listeners).forEach((id) => {
    const handler = eventData.listeners[id];
    if (isReadyForExecution(handler)) handler(event);
  });
};

// see https://github.com/microsoft/TypeScript/issues/32912#issuecomment-522142969
const DEFAULT_EVENT_OPTIONS: AddEventListenerOptions & EventListenerOptions = { passive: true };

const attachEmitterHandler = (
  eventType: EventType,
  eventData: EventData,
  eventSettings: EventSettings) => {
  if (eventData.listenersLength === 1) {
    eventSettings.emitter?.addEventListener(eventType, handleEmitterEvent, DEFAULT_EVENT_OPTIONS);
  }
};

const detachEmitterHandler = (
  eventType: EventType,
  eventData: EventData,
  eventSettings: EventSettings) => {
  if (eventData.listenersLength === 0) {
    eventSettings.emitter?.removeEventListener(eventType, handleEmitterEvent, DEFAULT_EVENT_OPTIONS);
  }
};

// ========================================================================================
// Public api
// ========================================================================================
const removeListener = (event:EventType, id: number) => {
  const eventData = getEventData(event);
  const eventSettings = getEventSettings(event);

  // protect from being called twice
  if (!eventData.listeners[id]) return;

  delete eventData.listeners[id];
  eventData.listenersLength -= 1;

  detachEmitterHandler(event, eventData, eventSettings);
};

const addListener = (event:EventType, callback: HandlerFunction) => {
  if (typeof callback !== 'function') throw new Error('Listener function required');
  if (!isDOMAvailable) return noop;

  const eventData = getEventData(event);
  const eventSettings = getEventSettings(event);

  eventData.lastIndex += 1;
  const id = eventData.lastIndex;
  const handler = eventSettings.wrapper
    ? eventSettings.wrapper(callback) as HandlerFunction : callback;
  handler._attachedAt = Date.now();

  eventData.listeners[id] = handler;
  eventData.listenersLength += 1;

  attachEmitterHandler(event, eventData, eventSettings);

  const removeEventListener = () => {
    if (typeof handler.cancel === 'function') {
      handler.cancel();
    }

    removeListener(event, id);
  };
  // Allow to empty the calls queue
  removeEventListener.cancel = handler.cancel;

  return removeEventListener;
};

if (isDOMAvailable) createEventSettings();

export default addListener;
